### ISSUE TITLE
[discussion] Native tensor-backed string array

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#101699]
### 🚀 The feature, motivation and pitch

This is useful to avoid copies related to copy-on-write problems with DataLoader: https://github.com/pytorch/pytorch/issues/13246

For fixed-size-characters (ascii: uint8 / utf_16: int16 / utf_32: int32) there is a prototype in https://gist.github.com/vadimkantorov/86c3a46bf25bed3ad45d043ae86fff57 , but some other designs can be considered. In essence, initially this means some fast parallelized APIs for conversion from python string lists and accessing individual elements / sub-lists.

This would be useful for some fast column-based string processing or mmap'ing some dataset files. Probably NumPy/HDF5/Apache Arrow / data frame libraries have also some support along these lines.

**Comment by vadimkantorov** [comment]
It also appears that ONNX / onnxrt supports some sort of string dtype and numpy-like string arrays:
- https://github.com/onnx/onnx/blob/main/docs/Operators-ml.md#aionnxmlarrayfeatureextractor - indexing operator for a string array
- https://github.com/onnx/onnx/blob/main/docs/Operators.md#StringNormalizer - basic string ops

**Comment by vadimkantorov** [comment]
in NumPy also string arrays do exist via the dtype `numpy.str_` https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.str_

This could be useful for more compat with NumPy too

**Comment by vadimkantorov** [comment]
A related discussion on importance of reducing GC pressure and number of Python objects:
https://github.com/pytorch/pytorch/pull/103339

So native StringArrays + some basic data frames are a useful idea!

**Comment by vadimkantorov** [comment]
Related: https://pytorch.org/rl/tensordict/saving.html#saving-memmory-mapped-tensordicts

**Comment by vmoens** [comment]
That would be an interesting feature to add to tensordict, but would need some discussion.
For usage in LLM it would defo be super useful. 

cc @tcbegley @apbard

**Comment by vadimkantorov** [comment]
@mikaylagawarecki btw maybe worth supporting arrow or expanding column-based mmap support as in https://twitter.com/rohanpaul_ai/status/1686472387091648512 ?

**Comment by dracifer** [comment]
> Going further, maybe some simplistic dataframe class can be added to PyTorch (being a tuple of tensors with having equal leftmost dim). These dataframes would primarily be used for some simple dataset serialization/deserialization / filtering and transformation. Ideally, a dataframe should support two modes of serialization: array-of-structs and column-based.

this idea matches pretty much our thoughts about adding "structure" to Pytorch to provide stronger types and memory layout and bridge data and model better with transformation libraries. cc @laurencer

**Comment by vadimkantorov** [comment]
some aspect is also padding / alignment by memory page size (which is OS/machine-dependent) for mmap sometimes

**Comment by vadimkantorov** [comment]
As one of forms of representation (padding-less), NestedTensors / JaggedLayout can be useful for these string arrays @cpuhrsch

**Comment by vadimkantorov** [comment]
Related: 
- https://github.com/triton-inference-server/pytorch_backend/pull/60

For model hosting, it's often useful to be able to store single strings in tensors and arrays of strings
